File: src/cudauvm/plugin-SiPixelClusterizer/SiPixelFedCablingMapGPUWrapperESProducer.cc
#include "CondFormats/SiPixelFedIds.h"
#include "CondFormats/SiPixelFedCablingMapGPU.h"
#include "CondFormats/SiPixelFedCablingMapGPUWrapper.h"
#include "Framework/ESProducer.h"
#include "Framework/EventSetup.h"
#include "Framework/ESPluginFactory.h"

#include <fstream>
#include <memory>

class SiPixelFedCablingMapGPUWrapperESProducer : public edm::ESProducer {
public:
  explicit SiPixelFedCablingMapGPUWrapperESProducer(std::filesystem::path const& datadir) : data_(datadir) {}
  void produce(edm::EventSetup& eventSetup);

private:
  std::filesystem::path data_;
};

void SiPixelFedCablingMapGPUWrapperESProducer::produce(edm::EventSetup& eventSetup) {
  {
    std::ifstream in(data_ / "fedIds.bin", std::ios::binary);
    in.exceptions(std::ifstream::badbit | std::ifstream::failbit | std::ifstream::eofbit);
    unsigned int nfeds;
    in.read(reinterpret_cast<char*>(&nfeds), sizeof(unsigned));
    std::vector<unsigned int> fedIds(nfeds);
    in.read(reinterpret_cast<char*>(fedIds.data()), sizeof(unsigned int) * nfeds);
    eventSetup.put(std::make_unique<SiPixelFedIds>(std::move(fedIds)));
  }
  {
    std::ifstream in(data_ / "cablingMap.bin", std::ios::binary);
    in.exceptions(std::ifstream::badbit | std::ifstream::failbit | std::ifstream::eofbit);
    SiPixelFedCablingMapGPU obj;
    in.read(reinterpret_cast<char*>(&obj), sizeof(SiPixelFedCablingMapGPU));
    unsigned int modToUnpDefSize;
    in.read(reinterpret_cast<char*>(&modToUnpDefSize), sizeof(unsigned int));
    std::vector<unsigned char> modToUnpDefault(modToUnpDefSize);
    in.read(reinterpret_cast<char*>(modToUnpDefault.data()), modToUnpDefSize);
    eventSetup.put(std::make_unique<SiPixelFedCablingMapGPUWrapper>(obj, std::move(modToUnpDefault)));
  }
}

DEFINE_FWK_EVENTSETUP_MODULE(SiPixelFedCablingMapGPUWrapperESProducer);

File: src/julia-serial/plugin-SiPixelClusterizer/SiPixelFedCablingMapGPUWrapperESProducer.jl
using .condFormatsSiPixelFedIds:SiPixelFedIds

using .recoLocalTrackerSiPixelClusterizerSiPixelFedCablingMapGPU:SiPixelFedCablingMapGPU
using .recoLocalTrackerSiPixelClusterizerSiPixelFedCablingMapGPU.pixelGPUDetails

using .recoLocalTrackerSiPixelClusterizerSiPixelFedCablingMapGPUWrapper:SiPixelFedCablingMapGPUWrapper

using .condFormatsSiPixelFedIds

using .ESPluginFactory


struct SiPixelFedCablingMapGPUWrapperESProducer <: ESProducer
    data::String  # Use String to represent the path

    function SiPixelFedCablingMapGPUWrapperESProducer(datadir::String)
        new(datadir)
    end
end

function readData(io::IOStream, type, size)
    container = Vector{type}(undef, size)
    return read!(io,container)
end


function readCablingMap(io::IOStream,es::EventSetup)

    maxSize = recoLocalTrackerSiPixelClusterizerSiPixelFedCablingMapGPU.pixelGPUDetails.MAX_SIZE

    fed = readData(io, UInt32, maxSize)
    link = readData(io, UInt32, maxSize)
    roc = readData(io, UInt32, maxSize)
    raw_id = readData(io, UInt32, maxSize)
    roc_in_det = readData(io, UInt32, maxSize)
    module_id = readData(io, UInt32, maxSize)
    bad_rocs = readData(io, UInt8, maxSize)
    size = UInt32(0)

    cablingMap = SiPixelFedCablingMapGPU(fed, link, roc, raw_id, roc_in_det, module_id, bad_rocs, size)

    mod_to_unp_def_size = read(io,UInt32)
    mod_to_unp_default = readData(io, UInt8, mod_to_unp_def_size)

    put!(es,SiPixelFedCablingMapGPUWrapper(cablingMap,mod_to_unp_default))
end


function produce(producer::SiPixelFedCablingMapGPUWrapperESProducer, eventSetup::EventSetup)
    fed_ids_file = joinpath(producer.data, "fedIds.bin")
    # Read fedIds.bin
    open(fed_ids_file, "r") do io
        nfeds = read(io, UInt32)
        fed_ids = Vector{UInt32}(undef, nfeds)
        read!(io, fed_ids)
        put!(eventSetup,SiPixelFedIds(fed_ids))
    end
    

    # Read cablingMap.bin
    cabling_map_file = joinpath(producer.data, "cablingMap.bin")
    
    open(cabling_map_file, "r") do io
        readCablingMap(io,eventSetup)
    end

end



add_plugin("SiPixelFedCablingMapGPUWrapperESProducer",x -> SiPixelFedCablingMapGPUWrapperESProducer(x))


